decayAmplitude/isobarDecayTopology.cc

```diff
@@ -543,126 +552,102 @@ isobarDecayTopology::getIsospinSymmetrization()
 
 	// Permutating all the particles and checking if the permutation makes sense
 	//
-	// First, loop over all the groups.
-	for(unsigned int i = 0; i < groups.size(); ++i) {
-		vector<unsigned int> group = groups.at(i);
-
-		// Again debug output
-		if(_debug) {
-			printDebug << "Group permutations " << i << endl;
+	const std::vector< std::vector< std::vector<unsigned int> > > permutations = isobarDecayTopologyHelper::getAllGroupPermutations( groups );
+	for( unsigned int i_permutation = 0; i_permutation < permutations.size(); ++i_permutation ){
+		const std::vector< std::vector<unsigned int> >& permutation = permutations[i_permutation];
+		std::vector<unsigned int> map( fsParts.size() );
+
+		for( unsigned int i_group = 0; i_group < groups.size(); ++i_group ){
+			for( unsigned int i_group_entry = 0; i_group_entry < groups[i_group].size(); ++i_group_entry){
+				map[ groups[i_group][i_group_entry] ] = permutation[i_group][i_group_entry];
+			}
 		}
-
-		// First we need a vector with the unity permutation, which will
-		// subsequently be permutated.
-		vector<unsigned int> permutations;
-		for(unsigned int j = 0; j < group.size(); ++j) {
-			permutations.push_back(j);
+		if(_debug) {
+			printDebug << "Processing permutation '";
+			for( unsigned int i = 0; i < map.size(); ++i ) cout << map[i];
+			cout << "'." << std::endl;
 		}
+		// Some first checks if the permutation makes sense.
+		bool breaking = false;
+		for(unsigned int j = 0; j < map.size(); ++j) {
 
-		// Now loop over all permutations for the current group.
-		do {
-			// Create a unity map. The map has one entry per final-state particle
-			// (unlike the group, which only holds the indices of the final-state
-			// particles in the group).
-			vector<unsigned int> map;
-			for(unsigned int j = 0; j < fsParts.size(); ++j) {
-				map.push_back(j);
-			}
+			if(j == map.at(j)) 	continue; // particle j was not swapped -> go on with check od next particle
 
-			// Now apply the permutation to the map. We get the index of the particle
-			// we want to swap from the group vector and the index of the particle we
-			// want to swap it with from the permutations vector. The whole thing is
-			// applied to the map vector.
-			for(unsigned int j = 0; j < group.size(); ++j) {
-				map.at(group.at(j)) = group.at(permutations.at(j));
+			// Make sure we are not swapping two indistinguishable particles.
+			if(fsParts.at(j)->name() == fsParts.at(map.at(j))->name()) {
```

Comment:
Maybe it would be worth replacing all the `.at()`s with `[]` while you are touching this part of the code.